### PR TITLE
James 1742

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPUrls.java [hunk 1]
```diff
@@ -23,4 +23,5 @@
 
     String JMAP = "/jmap";
     String AUTHENTICATION = "/authentication";
+    String DOWNLOAD = "/download/*";
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPUrls.java]: Are you sure the final star is part of the URL?

Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPUrls.java]: Following the spec: "The URL MUST contain a variable called blobId"

**Comment by chibenwa** [comment]
:+1:

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPUrls.java (hunk 1 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPUrls.java]: 1/ Right for the star
2/ It's not has easy with servlets: 
  we should add a star at the end of the URL which binds the servlet, 
  then we have to check inside the servlet that HttpServletRequest.getPathInfo() contains the blobId.

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.james.jmap.methods.integration.cucumber.MainStepdefs;
+import org.apache.james.mailbox.elasticsearch.EmbeddedElasticSearch;
+import org.apache.james.modules.CassandraJmapServerModule;
+import org.junit.rules.TemporaryFolder;
+
+import cucumber.api.java.After;
+import cucumber.api.java.Before;
+import cucumber.runtime.java.guice.ScenarioScoped;
+
+@ScenarioScoped
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java]: what does this annotation do ?

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        mainStepdefs.jmapServer = new GuiceJamesServer()
+                .combineWith(CassandraJamesServerMain.cassandraServerModule)
+                .overrideWith(new CassandraJmapServerModule(temporaryFolder, embeddedElasticSearch, cassandra));
+        mainStepdefs.awaitMethod = () -> embeddedElasticSearch.awaitForElasticSearch();
+        mainStepdefs.init();
+    }
+
+    @After
+    public void tearDown() {
+        mainStepdefs.tearDown();
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java]: any exception throw by on the two first methods will prevent tearDown to actually cleanup as much as possible.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    @When("^asking the download endpoint$")
+    public void getDownload() throws Exception {
+        if (userStepdefs.accessToken != null) {
+            response = getDownloadWithAuthorization();
+        } else {
+            response = getDownloadWithoutAuthorization();
+        }
+    }
+
+    private Response getDownloadWithoutAuthorization() {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: you failed at composing things : the two methods share 3  lines, why don't you init the builder this the common stuff then put the authorization header or not depending on accessToken availability ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature [hunk 5]
```diff
@@ -0,0 +1,21 @@
+Feature: Download authentication
+  As a James user
+  I want to access to the download endpoint
+
+  Background:
+    Given a domain named "domain.tld"
+
+  Scenario: A known user should access to the download endpoint
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature]: should have access

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    Given a current user with username "username@domain.tld" and password "secret"
+    When asking the download endpoint
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature]: what does "asking the download endpoint" means ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    Then the user should be authorized
+
+  Scenario: An known user should access to the download endpoint
+    Given an unknown user with username "unknown@domain.tld" and password "secret"
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature]: this one is not "current" ? it looks like a lack of consistency

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    When asking the download endpoint
+    Then the user should not be authorized
+
+  Scenario: A known user should not access to the download endpoint without a blobId
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature]: how is it related to "I want to access to the download endpoint" ?

File: server/protocols/jmap-integration-testing/memory-jmap-integration-testing/pom.xml [hunk 9]
```diff
@@ -182,7 +182,7 @@
                 </dependency>
                 <dependency>
                     <groupId>info.cukes</groupId>
-                    <artifactId>cucumber-picocontainer</artifactId>
+                    <artifactId>cucumber-guice</artifactId>
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/memory-jmap-integration-testing/pom.xml]: why this change ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java [hunk 10]
```diff
@@ -55,6 +55,11 @@ private JMAPServer(JMAPConfiguration jmapConfiguration,
                             .with(new AllowAllCrossOriginRequests(bypass(authenticationFilter).on("OPTIONS").only()))
                             .and(firstUserConnectionFilter)
                             .only()
+                        .serve(JMAPUrls.DOWNLOAD)
+                            .with(downloadServlet)
+                        .filter(JMAPUrls.DOWNLOAD)
+                            .with(new AllowAllCrossOriginRequests(bypass(authenticationFilter).on("OPTIONS").only()))
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java]: do we need to bypass authent on OPTIONS ? looks like it's not tested

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java [hunk 11]
```diff
@@ -35,26 +35,32 @@
 
 public class JMAPServer implements Configurable {
 
+    private static final String DOWNLOAD_URL = JMAPUrls.DOWNLOAD + "/*";
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java]: I would rather define another method on configurationBuilder that handle this `*` stuff

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java]: It acts as a `Singleton` for the scope of a scenario.

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java]: Done, the two classes `CassandraDownloadStepdefs` & `CassandraSetMailboxesMethodStepdefs` are also identical => refactored

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java (hunk 4 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/DownloadStepdefs.java]: Done

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature (hunk 5 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature]: ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature (hunk 6 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature (hunk 7 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature (hunk 8 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadAuthentication.feature]: Ok, renaming the feature.

File: server/protocols/jmap-integration-testing/memory-jmap-integration-testing/pom.xml (hunk 9 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/memory-jmap-integration-testing/pom.xml]: Picocontainer don't handle more complexe injection.

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java (hunk 11 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java]: OK

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java (hunk 2 above)
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java]: It's not the default behavior ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java (hunk 10 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/JMAPServer.java]: OK

File: server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/cassandra-jmap-integration-testing/src/test/java/org/apache/james/jmap/cassandra/cucumber/download/CassandraDownloadStepdefs.java]: No, the default behavior is to inject new instances each time.

**Comment by aduprat** [comment]
test this please

**Comment by aduprat** [comment]
#261
